src/compiler/binder.ts

```diff
@@ -864,7 +864,7 @@ function createBinder(): (file: SourceFile, options: CompilerOptions) => void {
                 return declareSymbol(container.symbol.exports!, container.symbol, node, symbolFlags, symbolExcludes);
             }
             else {
-                Debug.assertNode(container, canHaveLocals);
+                Debug.assertNode<Node, HasLocals>(container, canHaveLocals);

```

Comment:
These are all new as of #51682; it appears as though `canHaveLocals` does not behave very well under SFT, thanks to the same problems we have with other branded types. Without explicitly giving type arguments to `assertNode`, it's unable to figure out what to do here.

This is basically the same as why I had to change `tryCast`, except that `asserts arg is ...` _must_ have the output extend the input thanks to it being part of code flow. In `cast`/`tryCast`, I can cheat and break that relationship to some benefit, but only because the output has nothing to do with the input. Maybe that's reason enough to restore `cast`/`tryCast`, as annoying as that would make those functions behave.

Overall, this to me says that our inference needs improvement.